### DISCUSSION TITLE
SessionEvent listener not called if registered before `sessionmaker`

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#8466] [Usage Questions] [answered]
I was looking at: https://github.com/kvesteri/sqlalchemy-continuum/issues/275
And noticed that when listening to SessionEvents - the ordering of when a `sessionmaker` is used matters.

If I am listening to `sa.orm.session.Session` I need to ensure my `sessionmaker` is created BEFORE I listen to the event.
Otherwise the event listener is not getting called.

Is this the expected behavior ?
Or is there some way that the listening logic can be above the call to `sessionmaker` ?

Here is a working example:
```py
from sqlalchemy import Column, Integer, String, create_engine
from sqlalchemy.event import listen
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, mapper
from sqlalchemy.orm.session import Session

engine = create_engine('sqlite:////tmp/db.sqlite3')

session_1 = lambda session, flush_context, instances: print("IN session_1")
session_2 = lambda session, flush_context, instances: print("IN session_2")

factory = sessionmaker(bind=engine, autocommit=False)   # <--- Here both listeners are called
listen(Session, 'before_flush', session_1)
#                                                         <--- Moving the line here calls only session_2
listen(factory, 'before_flush', session_2)


Base = declarative_base()
class Article(Base):
    __tablename__ = 'article'
    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(255))

Base.metadata.create_all(engine)
session = factory()

article = Article(name='Some article')
session.add(article)
session.commit()
```

**Comment by AbdealiLoKo** [comment, marked as answer]
I was digging a bit deeper and found this calls both the listeners:
```py
listen(Session, 'before_flush', session_1)
factory = sessionmaker(bind=engine, autocommit=False)
listen(Session, 'before_flush', session_2)   # NOTE that this is on Session and not on factor
```

The issue seems to be very specific only when I have a listener on both: `Session` and `factory`
The listener on `Session` is ignored

I tried (naively) to change it to `listen(Session, 'before_flush', session_1, propagate=True)` but that didn't work either
(I has used propagate when listening to a Model 'before_insert' for something else)

I also tried changing how I create the session:
 - `session = Session(bind=engine)` - Calls session_1 only
 - `session = factory.class_(bind=engine)` - Same behavior as `factory()` - depending on where the `sessionmaker` was called - calls either session_2 only or both.

**Reply by zzzeek** [reply]
sure this is a bug, I don't have a fix for it right now but I will illustrate where it's happening as well as some agnostic test cases, moment

**Reply by zzzeek** [reply]
#8467
